client/client_test.go

```diff
@@ -112,6 +114,133 @@ func testNetworkMode(t *testing.T, sb integration.Sandbox) {
 	require.Contains(t, err.Error(), "network.host is not allowed")
 }
 
+func testFrontendImageNameTemplating(t *testing.T, sb integration.Sandbox) {
+	requiresLinux(t)
+	t.Parallel()
+	c, err := New(context.TODO(), sb.Address())
+	require.NoError(t, err)
+	defer c.Close()
+
+	registry, err := sb.NewRegistry()
+	if errors.Cause(err) == integration.ErrorRequirements {
+		t.Skip(err.Error())
+	}
+	require.NoError(t, err)
+
+	checkImageName := map[string]func(out, imageName string){
+		ExporterOCI: func(out, imageName string) {
+			// Nothing to check
+			return
+		},
+		ExporterDocker: func(out, imageName string) {
+			dt, err := ioutil.ReadFile(out)
+			require.NoError(t, err)
+
+			m, err := testutil.ReadTarToMap(dt, false)
+			require.NoError(t, err)
+
+			_, ok := m["oci-layout"]
+			require.True(t, ok)
+
+			var index ocispec.Index
+			err = json.Unmarshal(m["index.json"].Data, &index)
+			require.NoError(t, err)
+			require.Equal(t, 2, index.SchemaVersion)
+			require.Equal(t, 1, len(index.Manifests))
+
+			var dockerMfst []struct {
+				RepoTags []string
+			}
+			err = json.Unmarshal(m["manifest.json"].Data, &dockerMfst)
+			require.NoError(t, err)
+			require.Equal(t, 1, len(dockerMfst))
+			require.Equal(t, 1, len(dockerMfst[0].RepoTags))
+			require.Equal(t, "docker.io/library/"+imageName, dockerMfst[0].RepoTags[0])
+		},
+		ExporterImage: func(_, imageName string) {
+			// check if we can pull (requires containerd)
+			var cdAddress string
+			if cd, ok := sb.(interface {
+				ContainerdAddress() string
+			}); !ok {
+				return
+			} else {
+				cdAddress = cd.ContainerdAddress()
+			}
+
+			// TODO: make public pull helper function so this can be checked for standalone as well
+
+			client, err := containerd.New(cdAddress)
+			require.NoError(t, err)
+			defer client.Close()
+
+			ctx := namespaces.WithNamespace(context.Background(), "buildkit")
+
+			// check image in containerd
+			_, err = client.ImageService().Get(ctx, imageName)
+			require.NoError(t, err)
+
+			// deleting image should release all content
+			err = client.ImageService().Delete(ctx, imageName, images.SynchronousDelete())
+			require.NoError(t, err)
+
+			checkAllReleasable(t, c, sb, true)
+
+			_, err = client.Pull(ctx, imageName)
+			require.NoError(t, err)
+
+			err = client.ImageService().Delete(ctx, imageName, images.SynchronousDelete())
+			require.NoError(t, err)
+		},
+	}
+
+	// ctrl takes precedence over fe
+	for _, winner := range []string{"fe", "ctrl"} {
```

Comment:
maybe you can find better names for these. not easy to follow what is going on. Also better to use `t.Run()` with these loops.

Reply:
I'll improve the names and try a `t.Run` (the current layout was inspired by `testOCIExporter`)